### DISCUSSION TITLE
关于梯度下降的一些问题

### DISCUSSION THREAD

**Discussion by kkkkkkip** [fangwei123456/SpikingFlow#279] [General]
你好，我对snn的梯度下降有一些问题不大懂。在lif_fc_mnist中，损失函数是由发射脉冲频率和0/1的mse求得的，损失函数里的变量有几个呢，分别是什么？在梯度下降部分损失函数对什么求偏导呢，用链式法则对应的哪些变量呢？

**Comment by fangwei123456** [comment]
梯度计算部分，除了脉冲发放过程，其他部分与ANN是没有区别的

https://pytorch.org/docs/stable/notes/autograd.html

**Reply by kkkkkkip** [reply]
你好，看完了pytorch的文档，但是还是不大清楚spikingjelly的梯度下降是对哪些参数求导的，请问可以解释一下吗？

**Reply by fangwei123456** [reply]
前向传播完成后，按照前向传播的逆序，对所有`requires_grad=True`的tensor求导，非叶节点的`.grad`不会保存，用完即删；叶子节点的`.grad`存下，通常这些节点都是网络参数，在梯度下降时使用。

**Reply by kkkkkkip** [reply]
好的，谢谢您，对我的帮助很大

**Reply by kkkkkkip** [reply]
您好，根据代码理解，我的理解是一个epoch跑完T个timestep后得到的发射脉冲频率，频率和one_het_label求loss,再进行loss.backward的计算。因此，一个epoch只得到一个梯度的更新，进行一次群众更新。但我输出中间值得到每个timestep权重都进行更新了，但是不是T个timestep才得到一个loss吗？这点困惑很久了，希望能得到您的回复。

**Reply by fangwei123456** [reply]
不是一个epoch，而是一个iteration

**Reply by fangwei123456** [reply]
训练实际上是BPTT，建议找一些BPTT的资料看一下

**Reply by kkkkkkip** [reply]
[你好，还是不大清楚为什么在timestep=3的情况下，一个loss输出三个grad_output，可以解答一下吗](https://img-blog.csdnimg.cn/a5af2dc194aa4a55bf888a9e0ec3009e.png)

**Reply by fangwei123456** [reply]
`一个loss输出三个grad_output`指的是什么

**Reply by kkkkkkip** [reply]
loss = F.mse_loss(out_spikes_counter_frequency, label_one_hot)
loss.backward
torch.autograd.backward对mse自动求导微分输出的grad_output，同时也是surrogate里替代函数（例如piecewise_leaky_relu）的backward的输入

**Reply by fangwei123456** [reply]
T=3时发放率的计算涉及到3个脉冲tensor的相加，回传的自然是3个梯度

**Reply by kkkkkkip** [reply]
你好，请问这三个梯度具体计算过程是怎样呢

**Reply by fangwei123456** [reply]
和ANN没有区别。建议你尝试一下自己用torch的autograd function实现加减乘除以及矩阵乘法函数，以此熟悉pytorch的梯度计算流程。参考extend pytorch的教程

**Reply by kkkkkkip** [reply]
好的，但是关于脉冲发射频率的导数是怎么计算的呢，这一部分ann不涉及

**Reply by fangwei123456** [reply]
先累加脉冲：
https://github.com/fangwei123456/spikingjelly/blob/a75dd3b0b1d116ae89fcf795af19a13ab89be32c/spikingjelly/activation_based/examples/lif_fc_mnist.py#L166
然后除以总仿真步数：
https://github.com/fangwei123456/spikingjelly/blob/a75dd3b0b1d116ae89fcf795af19a13ab89be32c/spikingjelly/activation_based/examples/lif_fc_mnist.py#L167

**Reply by kkkkkkip** [reply]
嗯嗯，求频率的过程我知道，但是不是还需要对频率求导吗

**Reply by fangwei123456** [reply]
这些过程是pytorch的autograd完成的。这一部分与spikingjelly框架无关，是pytorch的核心功能
